### PR TITLE
Release 3.1.0

### PR DESCRIPTION
**Pull request by iloveeclipse**
Final 3.1.0 release, see #111.

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [x] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code
- [x] Added an entry into `gradlePlugin/CHANGELOG.md` if you have changed Gradle plugin code

### CODE REVIEW

**Comment by iloveeclipse** [comment]
@KengoTODA : similar to last time, the build fails due some version issues. Could you please check this?

**Comment by KengoTODA** [comment]
I hope rebuilding fixes this problem. Not sure why Gradle failed to download SpotBugs from Maven repository.

Please solve conflict, and use version `1.6.0` for gradle-plugin [to follow semver2](https://github.com/spotbugs/discuss/issues/12), then we can merge this. It is also good to wait #447, but it should not be mandatory.

**Comment by KengoTODA** [comment]
No idea for now. What I know is that, SpotBugs Gradle Plugin adds dependency dynamically like below,

https://github.com/spotbugs/spotbugs/blob/3.1.0_RC7/gradlePlugin/src/main/java/com/github/spotbugs/SpotBugsPlugin.java#L149-L156

Not sure this is really related with our build failure in this branch.

**Comment by iloveeclipse** [comment]
To be able to build I had to temporarily disable spotbugs task execution on spotbugs and spotbugs-annotations projects. I have no idea which logic is there but it seems that any NON SNAPSHOT version in the main build file causes gradle dependency resolution on gradlePlugin project going amok and require something which I can't understand. Please see the commit error message what I mean.

I think we can release with this temporary fix, but it would be great if someone with gradle knowledge would look at our build scripts to understand the root cause and may be even to provide a proper fix.

**Comment by iloveeclipse** [comment]
Arrrgh, the build for eclipse plugin must generate NOT com.github.spotbugs.plugin.eclipse_3.1.0.jar but com.github.spotbugs.plugin.eclipse_3.1.0.timestamp.jar, otherwise Eclipse will not update existing nightly builds with the release build!!!

**Comment by iloveeclipse** [comment]
> Arrrgh, the build for eclipse plugin must generate NOT com.github.spotbugs.plugin.eclipse_3.1.0.jar

Fixed with #473 , which must be merged before this one.

**Comment by iloveeclipse** [comment]
The build issue with spotbugs task and weird gradle error message is caused most likely by https://github.com/gradle/gradle/issues/2723

**Comment by KengoTODA** [comment]
If so, we can try to downgrade Gradle to 4.0.

And I think current change has conflict with #473, could you merge (or rebase to) latest `master` branch?

**Comment by KengoTODA** [comment]
I've confirmed that Gradle 4.0 cannot be solution: https://travis-ci.org/spotbugs/spotbugs/builds/292073473

**Comment by iloveeclipse** [comment]
I've rebased it now.

**Comment by KengoTODA** [comment]
I've added several changes to prepare for coming development.

We will tag rebased 94f4e9f3316c6082210856f8536789b1ae5814d9 as `4.0.0`, and create `release-3.1` branch from it. See https://github.com/spotbugs/discuss/issues/11 for detail.

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,8 @@ Currently the versioning policy of this project follows [Semantic Versioning v2.
 
 ## Unreleased - 2017-??-??
 
+## 3.1.0 - 2017-10-23
```
Comment (iloveeclipse) [inline CHANGELOG.md]: This date was too optimistic ... Should we change?

Comment (KengoTODA) [reply CHANGELOG.md]: OK I will.

**Comment by iloveeclipse** [comment]
> We will tag rebased 94f4e9f as 4.0.0

You mean probably 3.1.0?

**Comment by KengoTODA** [comment]
> You mean probably 3.1.0?

Oh, yes. Sorry for my mistake.

**Comment by iloveeclipse** [comment]
And probably now it will be 82b2bab

**Pull request merged**